### PR TITLE
👌 IMPROVE: Add dataclass serialisation to context

### PR DESCRIPTION
**Pull request by chrisjsewell**

### CODE REVIEW

File: tests/orm/utils/test_serialize.py [hunk 1]
```diff
@@ -162,3 +163,19 @@ def test_enum():
...
+
+
+def test_dataclass():
+    """Test serialization and deserialization of a ``dataclass``."""
+    obj = DataClass(1)
+    serialized = serialize.serialize(obj)
+    assert isinstance(serialized, str)
+
+    deserialized = serialize.deserialize_unsafe(serialized)
+    assert deserialized == DataClass(1)
```
Comment (sphuber) [inline tests/orm/utils/test_serialize.py:181]:
```suggestion
    assert deserialized == obj
```

File: aiida/orm/utils/serialize.py [hunk 2]
```diff
@@ -136,6 +158,8 @@ def represent_data(self, data):
             return represent_computer(self, data)
         if isinstance(data, orm.Group):
             return represent_group(self, data)
+        if is_dataclass(data) and not inspect.isclass(data):
```
Comment (sphuber) [inline aiida/orm/utils/serialize.py:161]: Just out of interest, why is the second clause necessary? What is an example of an object passing `is_dataclass(obj) and inspect.isclass(obj)`?

Comment (chrisjsewell) [reply aiida/orm/utils/serialize.py:161]:
```python
In [15]: import inspect
In [16]: from dataclasses import asdict, dataclass, is_dataclass

In [17]: @dataclass
    ...: class DataClass:
    ...:     """A dataclass for testing."""
    ...:     my_value: int
    ...: 

In [18]: is_dataclass(DataClass)  and inspect.isclass(DataClass)
Out[18]: True

In [19]: asdict(DataClass)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-19-c90f518d2aad> in <module>
----> 1 asdict(DataClass)

~/Documents/GitHub/aiida_core_develop/.tox/py38-pre-commit/lib/python3.8/dataclasses.py in asdict(obj, dict_factory)
   1070     """
   1071     if not _is_dataclass_instance(obj):
-> 1072         raise TypeError("asdict() should be called on dataclass instances")
   1073     return _asdict_inner(obj, dict_factory)
   1074 

TypeError: asdict() should be called on dataclass instances
```

Comment (chrisjsewell) [reply aiida/orm/utils/serialize.py:161]: for some reason, `_is_dataclass_instance` is private 🤷

**Review by sphuber — APPROVED** [review]
All good, thanks

**Pull request merged**